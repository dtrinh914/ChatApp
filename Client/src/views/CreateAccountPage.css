.CreateAccountPage{
    height: 100vh;
    display: flex;
    justify-content: center;
    align-items: center;
}
.CreateAccountPage form{
    display: flex;
    flex-direction: column;
}