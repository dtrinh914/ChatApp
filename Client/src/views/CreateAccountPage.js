import React from 'react';
import useInput from '../hooks/useInput';
import axios from 'axios';
import './CreateAccountPage.css'

function CreateAccountPage(){
    const [username, setUsername] = useInput();
    const [password, setPassword] = useInput();

    const handleSubmit = (e) =>{
        e.preventDefault()
        axios.post('/api/users/new', {
            username:username, 
            password:password, 
            withCredentials:true
        })
        .then( res => {
            console.log(res);
        })
    }
    
    return(
        <div className='CreateAccountPage' >
            <form onSubmit={handleSubmit}>
                <h1>Create an Account</h1>
                <label htmlFor="username">Username</label>
                <input type="text" name='username' id='username' value={username} onChange={setUsername} />
                <label htmlFor="password">Password</label>
                <input type="password" name='password' id='password' value={password} onChange={setPassword} />
                <button>Sign Up</button>
            </form>
        </div>
    );
}

export default CreateAccountPage; 