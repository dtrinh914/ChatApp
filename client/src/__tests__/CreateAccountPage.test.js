import React from 'react';
import CreateAccountPage from '../views/CreateAccountPage';
import {BrowserRouter} from 'react-router-dom';
import {cleanup, render, fireEvent, wait} from '@testing-library/react';
import axiosMock from 'axios';

const mockFunction = () => console.log('Triggered');

const component = <BrowserRouter><CreateAccountPage logInUser={mockFunction} /></BrowserRouter>;

afterEach(cleanup);

describe('CreateAccountPage component after each input', ()=>{
    afterEach(cleanup);

    it.each`
        username                     | result
        ${'test user'}               | ${'The username cannot contain any spaces.'}
        ${' test_user'}              | ${'The username cannot contain any spaces.'}
        ${' test_user '}             | ${'The username cannot contain any spaces.'}
        ${'a'}                       | ${'The username needs to be at least 3 characters.'}
        ${'reallylongtest_user_name'}| ${'The username cannot be longer than 15 characters.'}
    `('should return $result when username = $username', ({username, result})=>{
        const {getByTestId} = render(component);
        fireEvent.change(getByTestId('create-page-username-input'), {target:{value:username}});
        expect(getByTestId('create-page-username-helper').textContent).toBe(result);
    });

    it.each`
        password                                    | result
        ${'abc123'}                                 | ${'The password needs to be at least 8 characters.'}
        ${' abcd1234'}                              | ${'The password cannot contain any spaces.'}
        ${'abcd1234 '}                              | ${'The password cannot contain any spaces.'}
        ${'abc d123 4'}                             | ${'The password cannot contain any spaces.'}
        ${'thisisthelongestpasswordinthewholeworld'}| ${'The password cannot be more than 32 characters.'}
    `('should return $result when password = $password', ({password, result})=>{
        const {getByTestId} = render(component);
        fireEvent.change(getByTestId('create-page-password-input'), {target:{value:password}});
        expect(getByTestId('create-page-password-helper').textContent).toBe(result);
    });

    it('should have message "Passwords need to match."', ()=>{
        const {getByTestId} = render(component);
        fireEvent.change(getByTestId('create-page-password-input'), {target:{value:'abcd1234'}});
        fireEvent.change(getByTestId('create-page-confirm-input'), {target:{value:'1234abcd'}});

        expect(getByTestId('create-page-confirm-helper').textContent).toBe('Passwords need to match.');
    });
});

test('should create user and trigger log in function', async () =>{
    const {getByTestId, queryByTestId} = render(component);
    const spy = jest.spyOn(console, 'log');

    //loading bar should not be on page
    expect(queryByTestId('create-page-loading')).toBeNull();

    //fill out form
    fireEvent.change(getByTestId('create-page-username-input'), {target:{value:'test_user'}});
    fireEvent.change(getByTestId('create-page-password-input'), {target:{value:'1234abcd'}});
    fireEvent.change(getByTestId('create-page-confirm-input'),  {target:{value:'1234abcd'}});

    axiosMock.post.mockResolvedValueOnce({data:{status:1}});

    await wait(()=>{
        //press submit button
        fireEvent.click(getByTestId('create-page-submit'));
        //loading bar should appear
        expect(queryByTestId('create-page-loading')).toBeTruthy();
    });

    await wait(()=>{
        //loading bar should be gone 
        expect(queryByTestId('create-page-loading')).toBeNull();
        expect(spy).toHaveBeenCalled();
    })

});

test('should attempt to create user and return error message "This username has already been taken."', async () =>{
    const {getByTestId} = render(component);

    //fill out form
    fireEvent.change(getByTestId('create-page-username-input'), {target:{value:'test_user'}});
    fireEvent.change(getByTestId('create-page-password-input'), {target:{value:'1234abcd'}});
    fireEvent.change(getByTestId('create-page-confirm-input'),  {target:{value:'1234abcd'}});

    axiosMock.post.mockResolvedValueOnce({data:{status:0}});

    fireEvent.click(getByTestId('create-page-submit'));

    await wait(()=>{
        expect(getByTestId('create-page-username-helper').textContent).toBe('This username has already been taken.');
    })
});

