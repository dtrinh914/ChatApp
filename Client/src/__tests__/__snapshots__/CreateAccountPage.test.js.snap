// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="CreateAccountPage"
  >
    <form
      action="api/users/new"
      method="POST"
    >
      <h1>
        Create an Account
      </h1>
      <label
        for="username"
      >
        Username
      </label>
      <input
        id="username"
        name="username"
        type="text"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        type="password"
        value=""
      />
      <button>
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
